Pre-commit mypy run fails to catch and report type errors

Currently, `pip` uses `mypy` via a `pre-commit` [hook](https://github.com/pypa/pip/blob/master/.pre-commit-config.yaml#L71) and with some custom configuration specified in the `setup.cfg` [file](https://github.com/pypa/pip/blob/master/setup.cfg#L33).

However, due to a surprising difference in behavior in `mypy` between specifying individual files vs specifying directories to be checked, the way that `pre-commit` calls `mypy` ends up unexpectedly hiding 51 type errors that are present in the files being checked and that are prohibited by the `mypy` configuration specified in `setup.cfg`.

**Environment**

* pip version: Issue found on commit `8d042c448718355447f133c440f7a931a5ae57c0`, which is the tip of the `master` branch as of the time of this post.
* Python version: Python 3.8.5
* OS: Ubuntu 20.04.1 LTS (focal) inside WSL2
* mypy: version 0.790, same as specified in the `pre-commit` [hook config](https://github.com/pypa/pip/blob/master/.pre-commit-config.yaml#L72)

**Expected behavior**
The following two commands should produce equivalent results i.e. neither of them finds any type errors on the current `master` branch (exact commit hash specified above):
* `pre-commit run mypy --all-files`
* `mypy src/`

**Actual behavior & How to Reproduce**
Running `pre-commit run mypy --all-files` finds no type errors, as expected — this is enforced by CI.

However, running `mypy src/` finds 51 type errors. Adding the `--verbose` flag to the invocation confirms that `mypy` indeed found and applied the `setup.cfg` config, see this excerpt from the (extremely verbose) `mypy` log:
```
LOG:  Mypy Version:           0.790
LOG:  Config File:            setup.cfg
```

**Cause**
Some digging into `pre-commit` shows that it uses a slightly different but nevertheless reasonable `mypy` invocation: it explicitly lists all the files to be checked rather than specifying an entire directory. This is probably in order to be able to apply the `exclude` rules specified in the `pre-commit` [config](https://github.com/pypa/pip/blob/master/.pre-commit-config.yaml#L75). The abbreviated form of the `mypy` invocation it uses is the following (I include the full and rather long invocation at the bottom of the issue):
```
mypy --pretty < ... long list of space-separated .py files in the src/ directory ... >
```

Unlike the `mypy src/` invocation, this `mypy` invocation surprisingly finds no issues!

I have verified the following things:
* The `--pretty` flag has no effect on the output of the `mypy` invocation used by `pre-commit`.
* The files where `mypy` finds errors in the `mypy src/` invocation exist in the list of files in the invocation used by `pre-commit`.
* Both invocations successfully locate and use the `mypy` configuration specified in `setup.cfg` (confirmed by using the `--verbose` flag and checking that `mypy` reports `setup.cfg` as the source of the applied configuration).

Additionally, the type errors reported by `mypy src/` appear to be real issues. Just a couple of examples:
```
src/pip/_internal/resolution/resolvelib/factory.py:195: error: Item "None" of "Optional[Requirement]" has no attribute "name"
src/pip/_internal/resolution/resolvelib/factory.py:199: error: Item "None" of "Optional[Requirement]" has no attribute "specifier"
```
Here is the corresponding code: https://github.com/pypa/pip/blob/master/src/pip/_internal/resolution/resolvelib/factory.py#L179-L199

On line 195, `template` has type `InstallRequirement`, so `template.req` has type `Optional[Requirement]` per: https://github.com/pypa/pip/blob/master/src/pip/_internal/req/req_install.py#L95-L120

At no point is `template.req` checked for `None` before line 195 accesses `template.req.name`, so this is a real issue. Similarly, on line 199, `ireq.req` is also of type `Optional[Requirement]` but is never checked for `None` before `ireq.req.specifier` is accessed, so this is similarly a real issue.

One could argue that this surprising behavior is a bug in `mypy` and/or `pre-commit` and therefore should not be opened in the `pip` project. I will be opening a corresponding GitHub issue on `mypy` about this surprising behavior. However, I wanted to directly notify the `pip` maintainers of this problem as well, because currently the fact that `mypy` is running via `pre-commit` is in a way self-deceptive as the specified `mypy` configuration _does_ actually produce real errors when run outside of `pre-commit`.

If the maintainers would like me to, I'd be happy to help contribute to resolving this issue.

**Output**

Below is the full output of `mypy src/`. As mentioned above, I have verified that this output takes into account the `mypy` settings configured in `setup.cfg`.
```
$ mypy src/
src/pip/_internal/utils/misc.py:526: error: "None" has no attribute "require"src/pip/_internal/utils/logging.py:27: error: Incompatible types in assignment (expression has type "None", variable has type Module)src/pip/_internal/req/req_set.py:36: error: Argument 1 to "canonicalize_name" has incompatible type "Optional[str]"; expected "str"src/pip/_internal/req/req_set.py:44: error: Argument 1 to "canonicalize_name" has incompatible type "Optional[str]"; expected "str"src/pip/_internal/req/req_set.py:130: error: Item "None" of "Optional[Requirement]" has no attribute "specifier"src/pip/_internal/req/req_install.py:367: error: Incompatible types in assignment (expression has type "str", variable has type "NormalizedName")src/pip/_internal/network/auth.py:283: error: "Response" has no attribute "connection"src/pip/_internal/cli/progress_bars.py:21: error: Incompatible types in assignment (expression has type "None", variable has type Module)src/pip/_internal/cli/progress_bars.py:52: error: Argument 1 to "_select_progress_class" has incompatible type "Type[IncrementalBar]"; expected "Bar"src/pip/_internal/cli/progress_bars.py:52: error: Argument 2 to "_select_progress_class" has incompatible type "Type[Bar]"; expected "Bar"
src/pip/_internal/cli/progress_bars.py:125: error: Incompatible types in assignment (expression has type "Tuple[str, str, str]", base class "IncrementalBar" defined the type as "Tuple[str, str, str, str, str, str, str, str, str]")
src/pip/_internal/cli/progress_bars.py:191: error: "AnsiToWin32" has no attribute "isatty"
src/pip/_internal/cli/progress_bars.py:195: error: "AnsiToWin32" has no attribute "flush"
src/pip/_internal/cli/progress_bars.py:211: error: Cannot determine type of 'hide_cursor' in base class 'WindowsMixin'
src/pip/_internal/cli/progress_bars.py:211: error: Cannot determine type of 'file' in base class 'WindowsMixin'
src/pip/_internal/cli/progress_bars.py:215: error: Cannot determine type of 'hide_cursor' in base class 'WindowsMixin'
src/pip/_internal/cli/progress_bars.py:215: error: Cannot determine type of 'file' in base class 'WindowsMixin'
src/pip/_internal/cli/progress_bars.py:220: error: Cannot determine type of 'hide_cursor' in base class 'WindowsMixin'
src/pip/_internal/cli/progress_bars.py:220: error: Cannot determine type of 'file' in base class 'WindowsMixin'
src/pip/_internal/cli/progress_bars.py:225: error: Cannot determine type of 'hide_cursor' in base class 'WindowsMixin'
src/pip/_internal/cli/progress_bars.py:225: error: Cannot determine type of 'file' in base class 'WindowsMixin'
src/pip/_internal/cli/progress_bars.py:230: error: Cannot determine type of 'hide_cursor' in base class 'WindowsMixin'
src/pip/_internal/resolution/resolvelib/base.py:66: error: Argument 1 to "contains" of "SpecifierSet" has incompatible type "_BaseVersion"; expected "Union[Union[Version, LegacyVersion], str]"
src/pip/_internal/req/constructors.py:191: error: Incompatible types in assignment (expression has type "None", variable has type "Requirement")
src/pip/_internal/req/constructors.py:369: error: Incompatible types in assignment (expression has type "None", variable has type "Requirement")
src/pip/_internal/commands/show.py:65: error: "None" has no attribute "__iter__" (not iterable)
src/pip/_internal/commands/show.py:79: error: "None" has no attribute "__iter__" (not iterable)
src/pip/_internal/commands/debug.py:212: error: Argument 2 to "show_value" has incompatible type "bool"; expected "Optional[str]"
src/pip/_internal/wheel_builder.py:189: error: Argument 1 to "canonicalize_name" has incompatible type "Optional[str]"; expected "str"
src/pip/_internal/wheel_builder.py:205: error: Unsupported operand types for <= ("Version" and "None")
src/pip/_internal/wheel_builder.py:205: note: Left operand is of type "Optional[Version]"
src/pip/_internal/wheel_builder.py:261: error: Argument "backend" to "build_wheel_pep517" has incompatible type "Optional[Pep517HookCaller]"; expected "Pep517HookCaller"
src/pip/_internal/resolution/resolvelib/requirements.py:77: error: Item "None" of "Optional[Requirement]" has no attribute "name"
src/pip/_internal/resolution/resolvelib/requirements.py:111: error: Item "None" of "Optional[Requirement]" has no attribute "specifier"
src/pip/_internal/resolution/resolvelib/requirements.py:112: error: Argument 1 to "contains" of "SpecifierSet" has incompatible type "_BaseVersion"; expected "Union[Union[Version, LegacyVersion], str]"
src/pip/_internal/resolution/resolvelib/requirements.py:150: error: Argument 1 to "contains" of "SpecifierSet" has incompatible type "_BaseVersion"; expected "Union[Union[Version, LegacyVersion], str]"
src/pip/_internal/resolution/resolvelib/requirements.py:160: error: Argument 1 to "contains" of "SpecifierSet" has incompatible type "_BaseVersion"; expected "Union[Union[Version, LegacyVersion], str]"
src/pip/_internal/operations/check.py:52: error: Incompatible return value type (got "Tuple[Dict[NormalizedName, PackageDetails], bool]", expected "Tuple[Dict[str, PackageDetails], bool]")
src/pip/_internal/operations/check.py:139: error: Incompatible return value type (got "Set[NormalizedName]", expected "Set[str]")
src/pip/_internal/operations/check.py:139: note: Perhaps you need a type annotation for "installed"? Suggestion: "Set[str]"
src/pip/_internal/resolution/resolvelib/factory.py:195: error: Item "None" of "Optional[Requirement]" has no attribute "name"
src/pip/_internal/resolution/resolvelib/factory.py:199: error: Item "None" of "Optional[Requirement]" has no attribute "specifier"
src/pip/_internal/resolution/resolvelib/factory.py:364: error: No overload variant of "get" of "Mapping" matches argument type "str"
src/pip/_internal/resolution/resolvelib/factory.py:364: note: Possible overload variant:
src/pip/_internal/resolution/resolvelib/factory.py:364: note:     def get(self, key: NormalizedName) -> Optional[Distribution]
src/pip/_internal/resolution/resolvelib/factory.py:364: note:     <1 more non-matching overload not shown>
src/pip/_internal/resolution/resolvelib/resolver.py:31: error: Module 'pip._vendor.resolvelib.structs' has no attribute 'Graph'
src/pip/_internal/resolution/resolvelib/resolver.py:92: error: Argument 1 to "canonicalize_name" has incompatible type "Optional[str]"; expected "str"
src/pip/_internal/resolution/resolvelib/resolver.py:116: error: Incompatible types in assignment (expression has type "PipReporter", variable has type "PipDebuggingReporter")
src/pip/_internal/resolution/resolvelib/resolver.py:130: error: Item "None" of "Optional[Result]" has no attribute "mapping"
src/pip/_internal/commands/search.py:130: error: "None" has no attribute "__iter__" (not iterable)
src/pip/_internal/commands/list.py:212: error: "_BaseVersion" has no attribute "is_prerelease"
src/pip/_internal/commands/list.py:219: error: Incompatible return value type (got "None", expected "Distribution")
src/pip/_internal/commands/list.py:227: error: "Distribution" has no attribute "latest_version"; maybe "has_version" or "_get_version"?
src/pip/_internal/commands/list.py:228: error: "Distribution" has no attribute "latest_filetype"
src/pip/_internal/commands/install.py:53: error: Argument 1 to "canonicalize_name" has incompatible type "Optional[str]"; expected "str"
Found 51 errors in 18 files (checked 374 source files)
```

The full `mypy` invocation used by `pre-commit` and its output are the following:
```
$ mypy --pretty src/pip/_internal/commands/install.py src/pip/_internal/utils/glibc.py src/pip/_internal/models/candidate.py src/pip/_internal/commands/debug.py src/pip/_internal/req/req_tracker.py src/pip/_internal/cli/cmdoptions.py src/pip/_internal/commands/freeze.py src/pip/_internal/resolution/base.py src/pip/_internal/resolution/legacy/resolver.py src/pip/_internal/operations/install/legacy.py src/pip/_internal/commands/__init__.py src/pip/_internal/resolution/resolvelib/resolver.py src/pip/_internal/cli/status_codes.py src/pip/_internal/cli/main.py src/pip/_internal/operations/__init__.py src/pip/_internal/network/auth.py src/pip/_internal/req/req_file.py src/pip/_internal/models/search_scope.py src/pip/_internal/operations/prepare.py src/pip/_internal/network/download.py src/pip/_internal/vcs/git.py src/pip/_internal/utils/subprocess.py src/pip/_internal/req/req_install.py src/pip/_internal/utils/datetime.py src/pip/_internal/resolution/resolvelib/found_candidates.py src/pip/_internal/distributions/installed.py src/pip/_internal/utils/wheel.py src/pip/_internal/models/selection_prefs.py src/pip/_internal/vcs/subversion.py src/pip/_internal/utils/models.py src/pip/_internal/cli/main_parser.py src/pip/_internal/operations/install/wheel.py src/pip/_internal/utils/encoding.py src/pip/_internal/utils/inject_securetransport.py src/pip/_internal/utils/packaging.py src/pip/_internal/models/wheel.py src/pip/_internal/commands/list.py src/pip/_internal/cli/req_command.py src/pip/_internal/utils/compat.py src/pip/_internal/cli/spinners.py src/pip/_internal/req/req_uninstall.py setup.py src/pip/_internal/resolution/resolvelib/base.py src/pip/_internal/utils/logging.py src/pip/_internal/distributions/__init__.py src/pip/_internal/vcs/mercurial.py src/pip/_internal/pyproject.py src/pip/_internal/commands/show.py src/pip/_internal/utils/unpacking.py src/pip/_internal/utils/parallel.py src/pip/_internal/self_outdated_check.py src/pip/_internal/utils/hashes.py src/pip/_internal/utils/direct_url_helpers.py src/pip/_internal/operations/build/wheel.py src/pip/_internal/index/package_finder.py src/pip/_internal/distributions/wheel.py src/pip/_internal/network/utils.py src/pip/_internal/resolution/resolvelib/candidates.py src/pip/_internal/models/scheme.py src/pip/_internal/utils/__init__.py src/pip/_internal/operations/install/editable_legacy.py src/pip/_internal/network/__init__.py src/pip/_internal/operations/build/__init__.py src/pip/_internal/cli/parser.py src/pip/_internal/resolution/__init__.py src/pip/_internal/utils/entrypoints.py src/pip/_internal/distributions/sdist.py src/pip/_internal/models/target_python.py src/pip/_internal/cli/progress_bars.py tools/automation/release/__init__.py src/pip/_internal/vcs/__init__.py src/pip/_internal/operations/check.py src/pip/_internal/resolution/resolvelib/provider.py src/pip/_internal/commands/configuration.py src/pip/_internal/resolution/resolvelib/factory.py src/pip/_internal/__init__.py src/pip/_internal/distributions/base.py src/pip/_internal/vcs/versioncontrol.py src/pip/_internal/operations/build/metadata.py src/pip/_internal/utils/virtualenv.py src/pip/_internal/utils/pkg_resources.py src/pip/_internal/vcs/bazaar.py src/pip/__main__.py src/pip/_internal/resolution/resolvelib/reporter.py src/pip/_internal/operations/install/__init__.py src/pip/_internal/operations/freeze.py src/pip/_internal/index/__init__.py src/pip/_internal/operations/build/metadata_legacy.py src/pip/_internal/req/constructors.py src/pip/_internal/utils/typing.py src/pip/_internal/resolution/resolvelib/requirements.py src/pip/_internal/utils/deprecation.py tools/tox_pip.py src/pip/_internal/main.py src/pip/_internal/models/direct_url.py src/pip/_internal/cli/base_command.py src/pip/_internal/commands/help.py src/pip/_internal/resolution/resolvelib/__init__.py noxfile.py src/pip/_internal/utils/urls.py src/pip/_internal/network/session.py src/pip/_internal/commands/download.py src/pip/_internal/network/xmlrpc.py src/pip/_internal/commands/hash.py src/pip/_internal/cli/__init__.py src/pip/__init__.py src/pip/_internal/operations/build/wheel_legacy.py src/pip/_internal/resolution/legacy/__init__.py src/pip/_internal/configuration.py src/pip/_internal/cache.py src/pip/_internal/utils/misc.py src/pip/_internal/build_env.py src/pip/_internal/locations.py src/pip/_internal/commands/search.py src/pip/_internal/wheel_builder.py src/pip/_internal/utils/appdirs.py src/pip/_internal/network/lazy_wheel.py src/pip/_internal/utils/filetypes.py src/pip/_internal/models/format_control.py src/pip/_internal/exceptions.py src/pip/_internal/cli/command_context.py src/pip/_internal/models/__init__.py src/pip/_internal/cli/autocompletion.py src/pip/_internal/index/collector.py src/pip/_internal/models/index.py src/pip/_internal/commands/uninstall.py src/pip/_internal/utils/distutils_args.py src/pip/_internal/commands/check.py tools/automation/release/check_version.py src/pip/_internal/utils/filesystem.py src/pip/_internal/commands/cache.py src/pip/_internal/commands/completion.py src/pip/_internal/utils/compatibility_tags.py src/pip/_internal/commands/wheel.py src/pip/_internal/network/cache.py src/pip/_internal/utils/temp_dir.py src/pip/_internal/req/req_set.py src/pip/_internal/req/__init__.py src/pip/_internal/models/link.py src/pip/_internal/utils/setuptools_build.py
Success: no issues found in 140 source files
```

--- a/setup.cfg
+++ b/setup.cfg
@@ -31,14 +31,28 @@ per-file-ignores =
     tests/*: B011
 
 [mypy]
+follow_imports = silent
 ignore_missing_imports = True
 disallow_untyped_defs = True
 disallow_any_generics = True
 warn_unused_ignores = True
 
-[mypy-pip/_vendor/*]
-follow_imports = skip
+[mypy-pip._vendor.*]
 ignore_errors = True
+
+# These vendored libraries use runtime magic to populate things and don't sit
+# well with static typing out of the box. Eventually we should provide correct
+# typing information for their public interface and remove these configs.
+[mypy-pip._vendor.colorama]
+follow_imports = skip
+[mypy-pip._vendor.pkg_resources]
+follow_imports = skip
+[mypy-pip._vendor.progress.*]
+follow_imports = skip
+[mypy-pip._vendor.requests.*]
+follow_imports = skip
+[mypy-pip._vendor.retrying]
+follow_imports = skip
 
 [tool:pytest]
 addopts = --ignore src/pip/_vendor --ignore tests/tests_cache -r aR

--- a/setup.cfg
+++ b/setup.cfg
@@ -31,7 +31,6 @@ per-file-ignores =
     tests/*: B011
 
 [mypy]
-follow_imports = silent
 ignore_missing_imports = True
 disallow_untyped_defs = True
 disallow_any_generics = True

--- a/src/pip/_internal/commands/debug.py
+++ b/src/pip/_internal/commands/debug.py
@@ -4,7 +4,7 @@ import os
 import sys
 from optparse import Values
 from types import ModuleType
-from typing import Dict, List, Optional
+from typing import Any, Dict, List, Optional
 
 import pip._vendor
 from pip._vendor.certifi import where
@@ -24,7 +24,7 @@ logger = logging.getLogger(__name__)
 
 
 def show_value(name, value):
-    # type: (str, Optional[str]) -> None
+    # type: (str, Any) -> None
     logger.info('%s: %s', name, value)
 
 

--- a/src/pip/_internal/commands/install.py
+++ b/src/pip/_internal/commands/install.py
@@ -5,7 +5,7 @@ import os
 import shutil
 import site
 from optparse import SUPPRESS_HELP, Values
-from typing import TYPE_CHECKING, Iterable, List, Optional
+from typing import Iterable, List, Optional
 
 from pip._vendor.packaging.utils import canonicalize_name
 
@@ -22,7 +22,7 @@ from pip._internal.exceptions import CommandError, InstallationError
 from pip._internal.locations import get_scheme
 from pip._internal.metadata import get_environment
 from pip._internal.models.format_control import FormatControl
-from pip._internal.operations.check import check_install_conflicts
+from pip._internal.operations.check import ConflictDetails, check_install_conflicts
 from pip._internal.req import install_given_reqs
 from pip._internal.req.req_install import InstallRequirement
 from pip._internal.req.req_tracker import get_requirement_tracker
@@ -41,9 +41,6 @@ from pip._internal.wheel_builder import (
     build,
     should_build_for_install_command,
 )
-
-if TYPE_CHECKING:
-    from pip._internal.operations.check import ConflictDetails
 
 logger = logging.getLogger(__name__)
 

--- a/src/pip/_internal/commands/install.py
+++ b/src/pip/_internal/commands/install.py
@@ -5,7 +5,7 @@ import os
 import shutil
 import site
 from optparse import SUPPRESS_HELP, Values
-from typing import Iterable, List, Optional
+from typing import TYPE_CHECKING, Iterable, List, Optional
 
 from pip._vendor.packaging.utils import canonicalize_name
 
@@ -22,7 +22,7 @@ from pip._internal.exceptions import CommandError, InstallationError
 from pip._internal.locations import get_scheme
 from pip._internal.metadata import get_environment
 from pip._internal.models.format_control import FormatControl
-from pip._internal.operations.check import ConflictDetails, check_install_conflicts
+from pip._internal.operations.check import check_install_conflicts
 from pip._internal.req import install_given_reqs
 from pip._internal.req.req_install import InstallRequirement
 from pip._internal.req.req_tracker import get_requirement_tracker
@@ -42,6 +42,9 @@ from pip._internal.wheel_builder import (
     should_build_for_install_command,
 )
 
+if TYPE_CHECKING:
+    from pip._internal.operations.check import ConflictDetails
+
 logger = logging.getLogger(__name__)
 
 
@@ -49,7 +52,7 @@ def get_check_binary_allowed(format_control):
     # type: (FormatControl) -> BinaryAllowedPredicate
     def check_binary_allowed(req):
         # type: (InstallRequirement) -> bool
-        canonical_name = canonicalize_name(req.name)
+        canonical_name = canonicalize_name(req.name or "")
         allowed_formats = format_control.get_allowed_formats(canonical_name)
         return "binary" in allowed_formats
 

--- a/src/pip/_internal/metadata/base.py
+++ b/src/pip/_internal/metadata/base.py
@@ -1,10 +1,12 @@
 import logging
 import re
-from typing import Container, Iterator, List, Optional
+from typing import Container, Iterator, List, Optional, Union
 
-from pip._vendor.packaging.version import _BaseVersion
+from pip._vendor.packaging.version import LegacyVersion, Version
 
 from pip._internal.utils.misc import stdlib_pkgs  # TODO: Move definition here.
+
+DistributionVersion = Union[LegacyVersion, Version]
 
 logger = logging.getLogger(__name__)
 
@@ -34,7 +36,7 @@ class BaseDistribution:
 
     @property
     def version(self):
-        # type: () -> _BaseVersion
+        # type: () -> DistributionVersion
         raise NotImplementedError()
 
     @property

--- a/src/pip/_internal/metadata/pkg_resources.py
+++ b/src/pip/_internal/metadata/pkg_resources.py
@@ -3,14 +3,13 @@ from typing import Iterator, List, Optional
 
 from pip._vendor import pkg_resources
 from pip._vendor.packaging.utils import canonicalize_name
-from pip._vendor.packaging.version import _BaseVersion
 from pip._vendor.packaging.version import parse as parse_version
 
 from pip._internal.utils import misc  # TODO: Move definition here.
 from pip._internal.utils.packaging import get_installer
 from pip._internal.utils.wheel import pkg_resources_distribution_for_wheel
 
-from .base import BaseDistribution, BaseEnvironment
+from .base import BaseDistribution, BaseEnvironment, DistributionVersion
 
 
 class Distribution(BaseDistribution):
@@ -45,7 +44,7 @@ class Distribution(BaseDistribution):
 
     @property
     def version(self):
-        # type: () -> _BaseVersion
+        # type: () -> DistributionVersion
         return parse_version(self._dist.version)
 
     @property

--- a/src/pip/_internal/models/candidate.py
+++ b/src/pip/_internal/models/candidate.py
@@ -1,4 +1,3 @@
-from pip._vendor.packaging.version import _BaseVersion
 from pip._vendor.packaging.version import parse as parse_version
 
 from pip._internal.models.link import Link
@@ -14,7 +13,7 @@ class InstallationCandidate(KeyBasedCompareMixin):
     def __init__(self, name, version, link):
         # type: (str, str, Link) -> None
         self.name = name
-        self.version = parse_version(version)  # type: _BaseVersion
+        self.version = parse_version(version)
         self.link = link
 
         super().__init__(

--- a/src/pip/_internal/network/session.py
+++ b/src/pip/_internal/network/session.py
@@ -295,7 +295,7 @@ class PipSession(requests.Session):
             # Add a small amount of back off between failed requests in
             # order to prevent hammering the service.
             backoff_factor=0.25,
-        )
+        )  # type: ignore
 
         # Our Insecure HTTPAdapter disables HTTPS validation. It does not
         # support caching so we'll use it for all http:// URLs.

--- a/src/pip/_internal/operations/check.py
+++ b/src/pip/_internal/operations/check.py
@@ -15,18 +15,17 @@ from pip._internal.utils.misc import get_installed_distributions
 if TYPE_CHECKING:
     from pip._vendor.packaging.utils import NormalizedName
 
-    # Shorthands
-    PackageSet = Dict[NormalizedName, 'PackageDetails']
-    Missing = Tuple[str, Any]
-    Conflicting = Tuple[str, str, Any]
-
-    MissingDict = Dict[NormalizedName, List[Missing]]
-    ConflictingDict = Dict[NormalizedName, List[Conflicting]]
-    CheckResult = Tuple[MissingDict, ConflictingDict]
-    ConflictDetails = Tuple[PackageSet, CheckResult]
-
-
 logger = logging.getLogger(__name__)
+
+# Shorthands
+PackageSet = Dict['NormalizedName', 'PackageDetails']
+Missing = Tuple[str, Any]
+Conflicting = Tuple[str, str, Any]
+
+MissingDict = Dict['NormalizedName', List[Missing]]
+ConflictingDict = Dict['NormalizedName', List[Conflicting]]
+CheckResult = Tuple[MissingDict, ConflictingDict]
+ConflictDetails = Tuple[PackageSet, CheckResult]
 
 PackageDetails = namedtuple('PackageDetails', ['version', 'requires'])
 

--- a/src/pip/_internal/operations/check.py
+++ b/src/pip/_internal/operations/check.py
@@ -3,7 +3,7 @@
 
 import logging
 from collections import namedtuple
-from typing import Any, Callable, Dict, List, Optional, Set, Tuple
+from typing import TYPE_CHECKING, Any, Callable, Dict, List, Optional, Set, Tuple
 
 from pip._vendor.packaging.utils import canonicalize_name
 from pip._vendor.pkg_resources import RequirementParseError
@@ -12,23 +12,26 @@ from pip._internal.distributions import make_distribution_for_install_requiremen
 from pip._internal.req.req_install import InstallRequirement
 from pip._internal.utils.misc import get_installed_distributions
 
+if TYPE_CHECKING:
+    from pip._vendor.packaging.utils import NormalizedName
+
+    # Shorthands
+    PackageSet = Dict[NormalizedName, 'PackageDetails']
+    Missing = Tuple[str, Any]
+    Conflicting = Tuple[str, str, Any]
+
+    MissingDict = Dict[NormalizedName, List[Missing]]
+    ConflictingDict = Dict[NormalizedName, List[Conflicting]]
+    CheckResult = Tuple[MissingDict, ConflictingDict]
+    ConflictDetails = Tuple[PackageSet, CheckResult]
+
+
 logger = logging.getLogger(__name__)
-
-# Shorthands
-PackageSet = Dict[str, 'PackageDetails']
-Missing = Tuple[str, Any]
-Conflicting = Tuple[str, str, Any]
-
-MissingDict = Dict[str, List[Missing]]
-ConflictingDict = Dict[str, List[Conflicting]]
-CheckResult = Tuple[MissingDict, ConflictingDict]
-ConflictDetails = Tuple[PackageSet, CheckResult]
 
 PackageDetails = namedtuple('PackageDetails', ['version', 'requires'])
 
 
-def create_package_set_from_installed(**kwargs):
-    # type: (**Any) -> Tuple[PackageSet, bool]
+def create_package_set_from_installed(**kwargs: Any) -> Tuple["PackageSet", bool]:
     """Converts a list of distributions into a PackageSet.
     """
     # Default to using all packages installed on the system
@@ -59,7 +62,7 @@ def check_package_set(package_set, should_ignore=None):
     missing = {}
     conflicting = {}
 
-    for package_name in package_set:
+    for package_name, package_detail in package_set.items():
         # Info about dependencies of package_name
         missing_deps = set()  # type: Set[Missing]
         conflicting_deps = set()  # type: Set[Conflicting]
@@ -67,8 +70,8 @@ def check_package_set(package_set, should_ignore=None):
         if should_ignore and should_ignore(package_name):
             continue
 
-        for req in package_set[package_name].requires:
-            name = canonicalize_name(req.project_name)  # type: str
+        for req in package_detail.requires:
+            name = canonicalize_name(req.project_name)
 
             # Check if it's missing
             if name not in package_set:
@@ -114,7 +117,7 @@ def check_install_conflicts(to_install):
 
 
 def _simulate_installation_of(to_install, package_set):
-    # type: (List[InstallRequirement], PackageSet) -> Set[str]
+    # type: (List[InstallRequirement], PackageSet) -> Set[NormalizedName]
     """Computes the version of packages after installing to_install.
     """
 
@@ -136,7 +139,7 @@ def _simulate_installation_of(to_install, package_set):
 
 
 def _create_whitelist(would_be_installed, package_set):
-    # type: (Set[str], PackageSet) -> Set[str]
+    # type: (Set[NormalizedName], PackageSet) -> Set[NormalizedName]
     packages_affected = set(would_be_installed)
 
     for package_name in package_set:

--- a/src/pip/_internal/req/constructors.py
+++ b/src/pip/_internal/req/constructors.py
@@ -182,7 +182,7 @@ def parse_req_from_editable(editable_req):
 
     if name is not None:
         try:
-            req = Requirement(name)
+            req = Requirement(name)  # type: Optional[Requirement]
         except InvalidRequirement:
             raise InstallationError(f"Invalid requirement: '{name}'")
     else:
@@ -335,7 +335,7 @@ def parse_req_from_line(name, line_source):
             return text
         return f'{text} (from {line_source})'
 
-    if req_as_string is not None:
+    def _parse_req_string(req_as_string: str) -> Requirement:
         try:
             req = Requirement(req_as_string)
         except InvalidRequirement:
@@ -363,6 +363,10 @@ def parse_req_from_line(name, line_source):
                 if spec_str.endswith(']'):
                     msg = f"Extras after version '{spec_str}'."
                     raise InstallationError(msg)
+        return req
+
+    if req_as_string is not None:
+        req = _parse_req_string(req_as_string)  # type: Optional[Requirement]
     else:
         req = None
 

--- a/src/pip/_internal/req/req_install.py
+++ b/src/pip/_internal/req/req_install.py
@@ -349,7 +349,7 @@ class InstallRequirement:
 
         # When parallel builds are enabled, add a UUID to the build directory
         # name so multiple builds do not interfere with each other.
-        dir_name = canonicalize_name(self.name)
+        dir_name = canonicalize_name(self.name)  # type: str
         if parallel_builds:
             dir_name = f"{dir_name}_{uuid.uuid4().hex}"
 

--- a/src/pip/_internal/req/req_set.py
+++ b/src/pip/_internal/req/req_set.py
@@ -28,7 +28,7 @@ class RequirementSet:
         # type: () -> str
         requirements = sorted(
             (req for req in self.requirements.values() if not req.comes_from),
-            key=lambda req: canonicalize_name(req.name),
+            key=lambda req: canonicalize_name(req.name or ""),
         )
         return ' '.join(str(req.req) for req in requirements)
 
@@ -36,7 +36,7 @@ class RequirementSet:
         # type: () -> str
         requirements = sorted(
             self.requirements.values(),
-            key=lambda req: canonicalize_name(req.name),
+            key=lambda req: canonicalize_name(req.name or ""),
         )
 
         format_string = '<{classname} object; {count} requirement(s): {reqs}>'
@@ -122,6 +122,8 @@ class RequirementSet:
             existing_req and
             not existing_req.constraint and
             existing_req.extras == install_req.extras and
+            existing_req.req and
+            install_req.req and
             existing_req.req.specifier != install_req.req.specifier
         )
         if has_conflicting_requirement:

--- a/src/pip/_internal/resolution/resolvelib/base.py
+++ b/src/pip/_internal/resolution/resolvelib/base.py
@@ -1,14 +1,15 @@
-from typing import FrozenSet, Iterable, Optional, Tuple
+from typing import FrozenSet, Iterable, Optional, Tuple, Union
 
 from pip._vendor.packaging.specifiers import SpecifierSet
-from pip._vendor.packaging.utils import canonicalize_name
-from pip._vendor.packaging.version import _BaseVersion
+from pip._vendor.packaging.utils import NormalizedName, canonicalize_name
+from pip._vendor.packaging.version import LegacyVersion, Version
 
 from pip._internal.models.link import Link
 from pip._internal.req.req_install import InstallRequirement
 from pip._internal.utils.hashes import Hashes
 
 CandidateLookup = Tuple[Optional["Candidate"], Optional[InstallRequirement]]
+CandidateVersion = Union[LegacyVersion, Version]
 
 
 def format_name(project, extras):
@@ -62,7 +63,7 @@ class Constraint:
 class Requirement:
     @property
     def project_name(self):
-        # type: () -> str
+        # type: () -> NormalizedName
         """The "project name" of a requirement.
 
         This is different from ``name`` if this requirement contains extras,
@@ -97,7 +98,7 @@ class Requirement:
 class Candidate:
     @property
     def project_name(self):
-        # type: () -> str
+        # type: () -> NormalizedName
         """The "project name" of the candidate.
 
         This is different from ``name`` if this candidate contains extras,
@@ -118,7 +119,7 @@ class Candidate:
 
     @property
     def version(self):
-        # type: () -> _BaseVersion
+        # type: () -> CandidateVersion
         raise NotImplementedError("Override in subclass")
 
     @property

--- a/src/pip/_internal/resolution/resolvelib/candidates.py
+++ b/src/pip/_internal/resolution/resolvelib/candidates.py
@@ -1,10 +1,10 @@
 import logging
 import sys
-from typing import TYPE_CHECKING, Any, FrozenSet, Iterable, Optional, Tuple, Union
+from typing import TYPE_CHECKING, Any, FrozenSet, Iterable, Optional, Tuple, Union, cast
 
 from pip._vendor.packaging.specifiers import InvalidSpecifier, SpecifierSet
-from pip._vendor.packaging.utils import canonicalize_name
-from pip._vendor.packaging.version import Version, _BaseVersion
+from pip._vendor.packaging.utils import NormalizedName, canonicalize_name
+from pip._vendor.packaging.version import Version
 from pip._vendor.packaging.version import parse as parse_version
 from pip._vendor.pkg_resources import Distribution
 
@@ -19,7 +19,7 @@ from pip._internal.req.req_install import InstallRequirement
 from pip._internal.utils.misc import dist_is_editable, normalize_version_info
 from pip._internal.utils.packaging import get_requires_python
 
-from .base import Candidate, Requirement, format_name
+from .base import Candidate, CandidateVersion, Requirement, format_name
 
 if TYPE_CHECKING:
     from .factory import Factory
@@ -126,8 +126,8 @@ class _InstallRequirementBackedCandidate(Candidate):
         source_link,  # type: Link
         ireq,  # type: InstallRequirement
         factory,  # type: Factory
-        name=None,  # type: Optional[str]
-        version=None,  # type: Optional[_BaseVersion]
+        name=None,  # type: Optional[NormalizedName]
+        version=None,  # type: Optional[CandidateVersion]
     ):
         # type: (...) -> None
         self._link = link
@@ -166,7 +166,7 @@ class _InstallRequirementBackedCandidate(Candidate):
 
     @property
     def project_name(self):
-        # type: () -> str
+        # type: () -> NormalizedName
         """The normalised name of the project the candidate refers to"""
         if self._name is None:
             self._name = canonicalize_name(self.dist.project_name)
@@ -179,7 +179,7 @@ class _InstallRequirementBackedCandidate(Candidate):
 
     @property
     def version(self):
-        # type: () -> _BaseVersion
+        # type: () -> CandidateVersion
         if self._version is None:
             self._version = parse_version(self.dist.version)
         return self._version
@@ -262,8 +262,8 @@ class LinkCandidate(_InstallRequirementBackedCandidate):
         link,  # type: Link
         template,  # type: InstallRequirement
         factory,  # type: Factory
-        name=None,  # type: Optional[str]
-        version=None,  # type: Optional[_BaseVersion]
+        name=None,  # type: Optional[NormalizedName]
+        version=None,  # type: Optional[CandidateVersion]
     ):
         # type: (...) -> None
         source_link = link
@@ -315,8 +315,8 @@ class EditableCandidate(_InstallRequirementBackedCandidate):
         link,  # type: Link
         template,  # type: InstallRequirement
         factory,  # type: Factory
-        name=None,  # type: Optional[str]
-        version=None,  # type: Optional[_BaseVersion]
+        name=None,  # type: Optional[NormalizedName]
+        version=None,  # type: Optional[CandidateVersion]
     ):
         # type: (...) -> None
         super().__init__(
@@ -378,7 +378,7 @@ class AlreadyInstalledCandidate(Candidate):
 
     @property
     def project_name(self):
-        # type: () -> str
+        # type: () -> NormalizedName
         return canonicalize_name(self.dist.project_name)
 
     @property
@@ -388,7 +388,7 @@ class AlreadyInstalledCandidate(Candidate):
 
     @property
     def version(self):
-        # type: () -> _BaseVersion
+        # type: () -> CandidateVersion
         return parse_version(self.dist.version)
 
     @property
@@ -471,7 +471,7 @@ class ExtrasCandidate(Candidate):
 
     @property
     def project_name(self):
-        # type: () -> str
+        # type: () -> NormalizedName
         return self.base.project_name
 
     @property
@@ -482,7 +482,7 @@ class ExtrasCandidate(Candidate):
 
     @property
     def version(self):
-        # type: () -> _BaseVersion
+        # type: () -> CandidateVersion
         return self.base.version
 
     def format_for_error(self):
@@ -565,9 +565,9 @@ class RequiresPythonCandidate(Candidate):
 
     @property
     def project_name(self):
-        # type: () -> str
+        # type: () -> NormalizedName
         # Avoid conflicting with the PyPI package "Python".
-        return "<Python from Requires-Python>"
+        return cast(NormalizedName, "<Python from Requires-Python>")
 
     @property
     def name(self):
@@ -576,7 +576,7 @@ class RequiresPythonCandidate(Candidate):
 
     @property
     def version(self):
-        # type: () -> _BaseVersion
+        # type: () -> CandidateVersion
         return self._version
 
     def format_for_error(self):

--- a/src/pip/_internal/resolution/resolvelib/factory.py
+++ b/src/pip/_internal/resolution/resolvelib/factory.py
@@ -15,8 +15,7 @@ from typing import (
 )
 
 from pip._vendor.packaging.specifiers import SpecifierSet
-from pip._vendor.packaging.utils import canonicalize_name
-from pip._vendor.packaging.version import _BaseVersion
+from pip._vendor.packaging.utils import NormalizedName, canonicalize_name
 from pip._vendor.pkg_resources import Distribution
 from pip._vendor.resolvelib import ResolutionImpossible
 
@@ -44,7 +43,7 @@ from pip._internal.utils.misc import (
 )
 from pip._internal.utils.virtualenv import running_under_virtualenv
 
-from .base import Candidate, Constraint, Requirement
+from .base import Candidate, CandidateVersion, Constraint, Requirement
 from .candidates import (
     AlreadyInstalledCandidate,
     BaseCandidate,
@@ -139,8 +138,8 @@ class Factory:
         link,  # type: Link
         extras,  # type: FrozenSet[str]
         template,  # type: InstallRequirement
-        name,  # type: Optional[str]
-        version,  # type: Optional[_BaseVersion]
+        name,  # type: Optional[NormalizedName]
+        version,  # type: Optional[CandidateVersion]
     ):
         # type: (...) -> Optional[Candidate]
         # TODO: Check already installed candidate, and use it if the link and
@@ -202,10 +201,12 @@ class Factory:
         # all of them.
         # Hopefully the Project model can correct this mismatch in the future.
         template = ireqs[0]
+        assert template.req, "Candidates found on index must be PEP 508"
         name = canonicalize_name(template.req.name)
 
         extras = frozenset()  # type: FrozenSet[str]
         for ireq in ireqs:
+            assert ireq.req, "Candidates found on index must be PEP 508"
             specifier &= ireq.req.specifier
             hashes &= ireq.hashes(trust_internet=False)
             extras |= frozenset(ireq.extras)
@@ -368,7 +369,7 @@ class Factory:
     def get_dist_to_uninstall(self, candidate):
         # type: (Candidate) -> Optional[Distribution]
         # TODO: Are there more cases this needs to return True? Editable?
-        dist = self._installed_dists.get(candidate.name)
+        dist = self._installed_dists.get(candidate.project_name)
         if dist is None:  # Not installed, no uninstallation required.
             return None
 

--- a/src/pip/_internal/resolution/resolvelib/requirements.py
+++ b/src/pip/_internal/resolution/resolvelib/requirements.py
@@ -1,5 +1,5 @@
 from pip._vendor.packaging.specifiers import SpecifierSet
-from pip._vendor.packaging.utils import canonicalize_name
+from pip._vendor.packaging.utils import NormalizedName, canonicalize_name
 
 from pip._internal.req.req_install import InstallRequirement
 
@@ -24,7 +24,7 @@ class ExplicitRequirement(Requirement):
 
     @property
     def project_name(self):
-        # type: () -> str
+        # type: () -> NormalizedName
         # No need to canonicalise - the candidate did this
         return self.candidate.project_name
 
@@ -67,7 +67,8 @@ class SpecifierRequirement(Requirement):
 
     @property
     def project_name(self):
-        # type: () -> str
+        # type: () -> NormalizedName
+        assert self._ireq.req, "Specifier-backed ireq is always PEP 508"
         return canonicalize_name(self._ireq.req.name)
 
     @property
@@ -96,14 +97,14 @@ class SpecifierRequirement(Requirement):
 
     def is_satisfied_by(self, candidate):
         # type: (Candidate) -> bool
-        assert (
-            candidate.name == self.name
-        ), "Internal issue: Candidate is not for this requirement " " {} vs {}".format(
-            candidate.name, self.name
+        assert candidate.name == self.name, (
+            f"Internal issue: Candidate is not for this requirement "
+            f"{candidate.name} vs {self.name}"
         )
         # We can safely always allow prereleases here since PackageFinder
         # already implements the prerelease logic, and would have filtered out
         # prerelease candidates if the user does not expect them.
+        assert self._ireq.req, "Specifier-backed ireq is always PEP 508"
         spec = self._ireq.req.specifier
         return spec.contains(candidate.version, prereleases=True)
 
@@ -129,7 +130,7 @@ class RequiresPythonRequirement(Requirement):
 
     @property
     def project_name(self):
-        # type: () -> str
+        # type: () -> NormalizedName
         return self._candidate.project_name
 
     @property
@@ -160,7 +161,7 @@ class UnsatisfiableRequirement(Requirement):
     """A requirement that cannot be satisfied."""
 
     def __init__(self, name):
-        # type: (str) -> None
+        # type: (NormalizedName) -> None
         self._name = name
 
     def __str__(self):
@@ -176,7 +177,7 @@ class UnsatisfiableRequirement(Requirement):
 
     @property
     def project_name(self):
-        # type: () -> str
+        # type: () -> NormalizedName
         return self._name
 
     @property

--- a/src/pip/_internal/resolution/resolvelib/resolver.py
+++ b/src/pip/_internal/resolution/resolvelib/resolver.py
@@ -5,7 +5,7 @@ from typing import TYPE_CHECKING, Dict, List, Optional, Set, Tuple
 
 from pip._vendor.packaging.utils import canonicalize_name
 from pip._vendor.packaging.version import parse as parse_version
-from pip._vendor.resolvelib import ResolutionImpossible
+from pip._vendor.resolvelib import BaseReporter, ResolutionImpossible
 from pip._vendor.resolvelib import Resolver as RLResolver
 from pip._vendor.resolvelib.resolvers import Result
 
@@ -32,7 +32,7 @@ from .base import Constraint
 from .factory import Factory
 
 if TYPE_CHECKING:
-    from pip._vendor.resolvelib.structs import Graph
+    from pip._vendor.resolvelib.structs import DirectedGraph
 
 logger = logging.getLogger(__name__)
 
@@ -86,6 +86,7 @@ class Resolver(BaseResolver):
                     raise InstallationError(problem)
                 if not req.match_markers():
                     continue
+                assert req.name, "Constraint must be named"
                 name = canonicalize_name(req.name)
                 if name in constraints:
                     constraints[name] &= req
@@ -110,14 +111,14 @@ class Resolver(BaseResolver):
             user_requested=user_requested,
         )
         if "PIP_RESOLVER_DEBUG" in os.environ:
-            reporter = PipDebuggingReporter()
+            reporter = PipDebuggingReporter()  # type: BaseReporter
         else:
             reporter = PipReporter()
         resolver = RLResolver(provider, reporter)
 
         try:
             try_to_avoid_resolution_too_deep = 2000000
-            self._result = resolver.resolve(
+            result = self._result = resolver.resolve(
                 requirements, max_rounds=try_to_avoid_resolution_too_deep
             )
 
@@ -126,7 +127,7 @@ class Resolver(BaseResolver):
             raise error from e
 
         req_set = RequirementSet(check_supported_wheels=check_supported_wheels)
-        for candidate in self._result.mapping.values():
+        for candidate in result.mapping.values():
             ireq = candidate.get_install_requirement()
             if ireq is None:
                 continue
@@ -235,7 +236,7 @@ class Resolver(BaseResolver):
 
 
 def get_topological_weights(graph, expected_node_count):
-    # type: (Graph, int) -> Dict[Optional[str], int]
+    # type: (DirectedGraph, int) -> Dict[Optional[str], int]
     """Assign weights to each node based on how "deep" they are.
 
     This implementation may change at any point in the future without prior

--- a/src/pip/_internal/wheel_builder.py
+++ b/src/pip/_internal/wheel_builder.py
@@ -167,7 +167,7 @@ def _always_true(_):
 
 def _verify_one(req, wheel_path):
     # type: (InstallRequirement, str) -> None
-    canonical_name = canonicalize_name(req.name)
+    canonical_name = canonicalize_name(req.name or "")
     w = Wheel(os.path.basename(wheel_path))
     if canonicalize_name(w.name) != canonical_name:
         raise InvalidWheelFilename(
@@ -175,10 +175,11 @@ def _verify_one(req, wheel_path):
             "got {!r}".format(canonical_name, w.name),
         )
     dist = get_wheel_distribution(wheel_path, canonical_name)
-    if canonicalize_version(dist.version) != canonicalize_version(w.version):
+    dist_verstr = str(dist.version)
+    if canonicalize_version(dist_verstr) != canonicalize_version(w.version):
         raise InvalidWheelFilename(
             "Wheel has unexpected file name: expected {!r}, "
-            "got {!r}".format(str(dist.version), w.version),
+            "got {!r}".format(dist_verstr, w.version),
         )
     metadata_version_value = dist.metadata_version
     if metadata_version_value is None:
@@ -192,7 +193,7 @@ def _verify_one(req, wheel_path):
             and not isinstance(dist.version, Version)):
         raise UnsupportedWheel(
             "Metadata 1.2 mandates PEP 440 version, "
-            "but {!r} is not".format(str(dist.version))
+            "but {!r} is not".format(dist_verstr)
         )
 
 
@@ -242,6 +243,7 @@ def _build_one_inside_env(
         assert req.name
         if req.use_pep517:
             assert req.metadata_directory
+            assert req.pep517_backend
             wheel_path = build_wheel_pep517(
                 name=req.name,
                 backend=req.pep517_backend,

